### ISSUE TITLE
Building godot project causes C# editor plugins to reload and lose state

### ISSUE THREAD

**Issue by Beider** [godotengine/godot#51626]
### Godot version

3.3.2 stable mono win64

### System information

Windows 10

### Issue description

When you build your project in godot it causes all your editor plugins that you developed in C# to also reload, however new code is not active after rebuild. You still have to disable/enable the plugin to actually see changes.

So I am not sure what correct behavior should be here, I would guess either:

1. Building should reload plugins and any code changes should be applied to the EditorPlugin
2. Building should not reload editor plugins (probably preferred unless other issues like marshalling are fixed)

Personally I would lean towards option 2 because reloading causes you to lose a lot of state. In the current build of godot you also can't keep references to custom classes (unless they inherit from Godot.Object) inside your Node scripts. You have to move any such references into statics, then rather keep a key reference inside your Node scripts to look them up. If you do not you will get marshalling issues (which there already exists bug reports for).

```csharp
// Example workaround for marshalling

// Static reference to custom class
private static Dictionary<string, CustomClass> CustomClassLookup= new Dictionary<string, CustomClass>();

// List containing keys for lookup in dicitonary
private List<string> KeyList = new List<string>();
```

In addition all static data you keep will get lost on reload. The upside to this is that you can use this to easily detect if your C# EditorPlugin has been reloaded by creating a static bool and then checking and flipping it during the _Process(). If the boolean resets then your plugin has been reloaded as the value of the static is not kept on rebuild.

```csharp
// Code to detect rebuild
public static bool Rebuild = true;
public override void _Process(float delta)
{
	if (Rebuild)
	{
		GD.Print("EditorPlugin was reloaded");
		Rebuild = false;
	}
}

### Steps to reproduce

1. Create a C# EditorPlugin for godot that contains the rebuild detection code provided above.
2. Load the editor plugin in some project that contain C# code.
3. Change some C# code in your project and rebuild the project

### Minimal reproduction project

_No response_

**Comment by Calinou** [comment]
Possible duplicate of https://github.com/godotengine/godot/issues/44147?

**Comment by neikeq** [comment]
I got a bit confused so I'm not sure if this is a bug or not. Does the new code not run after rebuild? If so can you give a bit more details about that?

Regarding the plugins being reloaded when rebuilding, this is by design. The entire domain needs to be reloaded to reflect any changes. In the future with .NET 6 we will be able to use AssemblyLoadContext instead, so perhaps we can design it to load plugins into their own contexts and only reload the ones that change.
As to how to deal with this right now, you can look at how things work in the built-in editor [code](https://github.com/godotengine/godot/tree/03e2544d50b73a56907694fc45cb82ad11c74a95/modules/mono/editor/GodotTools/GodotTools). Since serialization is limited, you can use the `ISerializationListener` interface to listen for `OnBeforeSerialize` and `OnAfterDeserialize` to try to workaround problems. It has the same purpose as Unity's [ISerializationCallbackReceiver](https://docs.unity3d.com/ScriptReference/ISerializationCallbackReceiver.html).

**Comment by Beider** [comment]
> I got a bit confused so I'm not sure if this is a bug or not. Does the new code not run after rebuild? If so can you give a bit more details about that?

That is correct, if you make changes to a C# plugin the new code you add does not appear to run on reload. This means you have to go and disable/enable the plugin again to get the new code to run. At least this is what i observed while developing my addon.

Thank you for the hints on Serialization and also I think if it will be possible in the future to not reload plugins that have not been changed that would be preferable. At least that would mean that you don't need to deal with serialization for your plugin if you do not want to.

**Comment by Zerbu** [comment]
This issue still exists in 2026. It took me **hours** to figure out why my editor code kept bugging out. Turns out, this was the cause. I think something needs to be done, as this makes making C# plugins really frustrating.

The engine appears to reconstruct each object and copy the data over. Any code in the constructor is rerun, but _EnterTree and _Ready are not. This means if you use the constructor to create child nodes, you'll end up with duplicates. Primitives and Godot objects seem to be copied over fine, but native C# objects are lost. This means if you don't want your plugin to break on recompile, you can't use any class that doesn't inherit from GodotObject. This includes objects from the standard .NET libraries.
